dev/AppNotifications/ShellLocalization.cpp

```diff
@@ -0,0 +1,265 @@
+    return std::wstring(localFolderPath.c_str());
+}
+
+inline wil::unique_hicon RetrieveIconFromProcess()
+{
+    std::wstring processPath{};
+    THROW_IF_FAILED(wil::GetModuleFileNameExW(GetCurrentProcess(), nullptr, processPath));
+
+    // Extract the small icon from the first .ico, if failed extract the large icon.
+    // Small icon is good enough for an App Notification icon since higher quality doesn't really impact in a substantial way.
+    wil::unique_hicon hIcon{};
+    if (!ExtractIconExW(processPath.c_str(), 0 /* index */, nullptr /* Large icon */, &hIcon, 1))
+    {
+        THROW_HR_IF(E_FAIL, ExtractIconExW(processPath.c_str(), 0, &hIcon, nullptr /* Small Icon */, 1) == 0);
+    }
+
+    return hIcon;
+}
+
+HRESULT Microsoft::Windows::AppNotifications::ShellLocalization::RetrieveAssetsFromProcess(_Out_ Microsoft::Windows::AppNotifications::ShellLocalization::AppNotificationAssets& assets) noexcept try
+{
+    wil::unique_hicon hIcon{ RetrieveIconFromProcess() };
+
+    std::wstring notificationAppId{ Microsoft::Windows::AppNotifications::Helpers::RetrieveNotificationAppId() };
+
+    // path: C:\Users\<currentUser>\AppData\Local\Microsoft\WindowsAppSDK\{AppGUID}.png
+    std::wstring iconFilePath{ RetrieveLocalFolderPath().c_str() + c_backSlash + notificationAppId + c_pngExtension };
+    WriteHIconToPngFile(hIcon, iconFilePath.c_str());
+
+    assets.displayName = Microsoft::Windows::AppNotifications::Helpers::GetDisplayNameBasedOnProcessName();
+    assets.iconFilePath = iconFilePath;
+
+    // TODO: Clean up the icon file in case UnregisterAll wasn't called.
+
+    return S_OK;
+}
+CATCH_RETURN()
+
+// Do nothing. This is just a placeholder while the UDK is ingested with the proper API.
+HRESULT Microsoft::Windows::AppNotifications::ShellLocalization::RetrieveAssetsFromShortcut(_Out_ Microsoft::Windows::AppNotifications::ShellLocalization::AppNotificationAssets& /*assets*/)
+{
+    // THROW_HR_IF_MSG(E_UNEXPECTED, IsIconFileExtensionSupported(iconFilePath));
+
+    return E_NOTIMPL;
+}
+
+HRESULT Microsoft::Windows::AppNotifications::ShellLocalization::RetrieveDefaultAssets(_Out_ Microsoft::Windows::AppNotifications::ShellLocalization::AppNotificationAssets& assets) noexcept try
+{
+    assets.displayName = Microsoft::Windows::AppNotifications::Helpers::GetDisplayNameBasedOnProcessName();
+    assets.iconFilePath = L""; // No icon filepath for now. This is coming in a future FrameworkUdk ingestion.
+
+    return S_OK;
+}
+CATCH_RETURN()
+
+void Microsoft::Windows::AppNotifications::ShellLocalization::DeleteIconFromCache() noexcept try
+{
+    std::wstring notificationAppId{ Microsoft::Windows::AppNotifications::Helpers::RetrieveNotificationAppId() };
+
+    // path: C:\Users\<currentUser>\AppData\Local\Microsoft\WindowsAppSDK\{AppGUID}.png
+    std::wstring iconFilePath{ RetrieveLocalFolderPath().c_str() + c_backSlash + notificationAppId + c_pngExtension };
+
+    THROW_IF_WIN32_BOOL_FALSE(DeleteFile(iconFilePath.c_str()));
```

Comment:
For the first time, I am going to have to say that this needs to be best effort. ie. If the file is in use or of the file is absent for some reason, please ignore. Thoughts?

Reply:
But we should LOG_IF_FALSE for sure.

Reply:
Also it may be better to log why the delete failed. You will need to look for the win32 error:https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-deletefilea

Reply:
Actually, I'm using the API wrongly. It returns non-zero if successful, otherwise it's a failure and should use GetLastError to know what the issue is: https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-deletefilew

Reply:
I would say, let DeleteFile to throw and let the caller decide if propagate the error or log it.

Reply:
Then why don't you check if file exists before you call delete? Then caller can probably treat this as a genuine error.